3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/VeroAdapter.scala

```diff
@@ -0,0 +1,167 @@
+ */
+object VeroAdapter extends Adapter {
+
+  // Vendor name for Failure Message
+  private val VendorName = "Vero"
+
+  // Expected content type for a request body
+  private val ContentType = "application/json"
+
+  // Tracker version for an Vero webhook
+  private val TrackerVersion = "com.getvero-v1"
+
+  // Schemas for reverse-engineering a Snowplow unstructured event
+  private val EventSchemaMap = Map(
+    "bounced"      -> SchemaKey("com.getvero", "bounced", "jsonschema", "1-0-0").toSchemaUri,
+    "clicked"      -> SchemaKey("com.getvero", "clicked", "jsonschema", "1-0-0").toSchemaUri,
+    "delivered"    -> SchemaKey("com.getvero", "delivered", "jsonschema", "1-0-0").toSchemaUri,
+    "opened"       -> SchemaKey("com.getvero", "opened", "jsonschema", "1-0-0").toSchemaUri,
+    "sent"         -> SchemaKey("com.getvero", "sent", "jsonschema", "1-0-0").toSchemaUri,
+    "unsubscribed" -> SchemaKey("com.getvero", "unsubscribed", "jsonschema", "1-0-0").toSchemaUri,
+    "user_created" -> SchemaKey("com.getvero", "created", "jsonschema", "1-0-0").toSchemaUri,
+    "user_updated" -> SchemaKey("com.getvero", "updated", "jsonschema", "1-0-0").toSchemaUri
+  )
+
+  /**
+   * Converts a payload into a single validated event
+   * Expects a valid json returns failure if one is not present
+   * For "user_updated" events the field is named "action" instead of "type"
+   * in which case the event type will be set to Some("user_updated")
+   *
+   * @param body_json Payload body that is sent by Vero
+   * @param payload The details of the payload
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+  private def payloadBodyToEvent(body_json: String, payload: CollectorPayload): Validated[RawEvent] =
+    try {
+      val parsed    = parse(body_json)
+      val eventType = (parsed \ "type").extractOpt[String].orElse(Some("user_updated"))
+      val formattedEvent = eventType match {
+        case Some("user_updated") => reformatParameters(parsed)
+        case _ => {
+          cleanupJsonEventValues(
+            parsed,
+            ("type", eventType.get).some,
+            eventType.get.concat("_at")
+          )
+        }
+      }
+
+      val reformattedEvent = reformatParameters(formattedEvent)
+
+      lookupSchema(eventType, VendorName, EventSchemaMap) map { schema =>
+        RawEvent(
+          api = payload.api,
+          parameters = toUnstructEventParams(
+            TrackerVersion,
+            toMap(payload.querystring),
+            schema,
+            reformattedEvent,
+            "srv"
+          ),
+          contentType = payload.contentType,
+          source      = payload.source,
+          context     = payload.context
+        )
+      }
+
+    } catch {
+      case e: JsonParseException => {
+        val exception = JU.stripInstanceEtc(e.toString).orNull
+        s"$VendorName event failed to parse into JSON: [$exception]".failureNel
+      }
+    }
```

Comment:
I don't think we need to wrap the whole thing in a try/catch but just `parse`? or am I wrong (cc @chuwy)

Reply:
@chuwy made a similar comment I THINK in the Marketo adapter but I'm not sure what needs to be changed

Reply:
```scala
val parsed = Try(parse(bodyJson)
val eventType = ...
```

instead of

```scala
try {
  val parsed = parse(bodyJson)
  val eventType = ...
} catch { ... }
```

Reply:
Hey @BenFradet, do we have a function in SCE codebase that parses string into `Validation[Something, JValue]`, just to show @misterpig how ideally it would look like.